imap_processing/spice/spin.py

```diff
@@ -0,0 +1,233 @@
+    query_met_times = np.asarray(query_met_times)
+    is_scalar = query_met_times.ndim == 0
+    if is_scalar:
+        # Force scalar to array because np.asarray() will not
+        # convert scalar to array
+        query_met_times = np.atleast_1d(query_met_times)
+
+    # Make sure input times are within the bounds of spin data
+    spin_df_start_time = spin_df["spin_start_time"].values[0]
+    spin_df_end_time = (
+        spin_df["spin_start_time"].values[-1] + spin_df["spin_period_sec"].values[-1]
+    )
+    input_start_time = query_met_times.min()
+    input_end_time = query_met_times.max()
+    if input_start_time < spin_df_start_time or input_end_time >= spin_df_end_time:
+        raise ValueError(
+            f"Query times, {query_met_times} are outside of the spin data range, "
+            f"{spin_df_start_time, spin_df_end_time}."
+        )
+
+    # Find all spin time that are less or equal to query_met_times.
+    # To do that, use side right, a[i-1] <= v < a[i], in the searchsorted.
+    # Eg.
+    # >>> df['a']
+    # array([0, 15, 30, 45, 60])
+    # >>> np.searchsorted(df['a'], [0, 13, 15, 32, 70], side='right')
+    # array([1, 1, 2, 3, 5])
+    last_spin_indices = (
+        np.searchsorted(spin_df["spin_start_time"], query_met_times, side="right") - 1
+    )
+    # Generate a dataframe with one row per query time
+    out_df = spin_df.iloc[last_spin_indices]
+
+    # Calculate spin phase
+    spin_phases = (query_met_times - out_df["spin_start_time"].values) / out_df[
+        "spin_period_sec"
+    ].values
+
+    # Check for invalid spin phase using below checks:
+    # 1. Check that the spin phase is in valid range, [0, 1).
+    # 2. Check invalid spin phase using spin_phase_valid,
+    #   spin_period_valid columns.
+    invalid_spin_phase_range = (spin_phases < 0) | (spin_phases >= 1)
+
+    invalid_spins = (out_df["spin_phase_valid"].values == 0) | (
+        out_df["spin_period_valid"].values == 0
+    )
+    bad_spin_phases = invalid_spin_phase_range | invalid_spins
+    spin_phases[bad_spin_phases] = np.nan
+
+    # Add spin_phase column to output dataframe
+    out_df["sc_spin_phase"] = spin_phases
+
+    return out_df
+
+
+def get_spin_angle(
+    spin_phases: Union[float, npt.NDArray],
+    degrees: bool = False,
+) -> Union[float, npt.NDArray]:
+    """
+    Convert spin_phases to radians or degrees.
+
+    Parameters
+    ----------
+    spin_phases : float or np.ndarray
+        Instrument or spacecraft spin phases. Spin phase is a
+        floating point number in the range [0, 1) corresponding to the
+        spin angle / 360.
+    degrees : bool
+        If degrees parameter is True, return angle in degrees otherwise return angle in
+        radians. Default is False.
+
+    Returns
+    -------
+    spin_phases : float or np.ndarray
+        Spin angle in degrees or radians for the input query times.
+    """
+    if np.any(spin_phases < 0) or np.any(spin_phases > 1):
```

Comment:
Looking at above code, should we change this to this?
```suggestion
    if np.any(spin_phases < 0) or np.any(spin_phases >= 1):
```